### ISSUE TITLE
transformer example polish

### ISSUE THREAD

**Issue by ZhitingHu** [asyml/texar-pytorch#66]
- [x] `texar.modules.Transformer` in its current form is far from being qualified as a Texar module. Please move it to under examples/

- [x] Can we use Texar data modules instead of torchtext which is not as clean as ours, e.g., 
https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L185-L192

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L43-L46
The explanation "Either train_and_evaluate or test" is inconsistent to the code, which allows "train_and_evaluate"/"eval"/"test"....   Also, call it either "eval" or "evaluate", not both

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L73
avoid using uncommon APIs like `getattr`, which harms readability. Is it necessary to use `getattr` here?

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L82-L86
Is it necessary for all models? Can we simply this? E.g., add a `maybe_to_cuda()` method to `texar.ModuleBase` 

- [ ] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L97-L99 
Can this be replaced with Texar APIs? Quite a few hyperparameters hard-coded here

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L102
Is it common / necessary ?

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L114
Is it common / necessary ?

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L137
If it's only for temporary file, use python `tempfile`

**Comment by ZhitingHu** [comment]
For `transformer.py`, which is supposed to be in 

- [ ] Add copyright header!

- [ ] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/texar/modules/transformer.py#L11-L16
do not expose the internal details. Use only, e.g., `from texar.modules import TransformerDecoder`

- [ ] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/texar/modules/transformer.py#L29-L30
vocab has been loaded once, do not load again

- [ ] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/texar/modules/transformer.py#L33-L34
Avoid magic constant 

(Quite a few other issues, will review after it's moved to examples/ and polished)

**Comment by huzecong** [comment]
>   https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L185-L192
>  Can we use Texar data modules instead of torchtext which is not as clean as ours, e.g.,

`torchtext` was required here because dynamic-sized batches (that depends on the total number of tokens in the batch) are required. Currently the data module does not support this, but such functionality can be easily added (see issue #43).

> https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L73
>   
>   avoid using uncommon APIs like `getattr`, which harms readability. Is it necessary to use `getattr` here?

I believe this was here because `config_model` did not contain the `beam_width` variable before beam search was implemented, so `getattr` with a default value was required. Now we can safely remove this.

> https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L82-L86
>   
>   Is it necessary for all models? Can we simply this? E.g., add a `maybe_to_cuda()` method to `texar.ModuleBase`

This is necessary, but what I would do is to check whether CUDA is available, and then create a `device` variable of type `torch.device`, and simply call `model = model.to(device)`, i.e.:
```python
if torch.cuda.is_available():
	device = torch.cuda.current_device()
else:
	device = 'cpu'
device = torch.device(device)
```
It's not sufficient to add `maybe_to_cuda()` to `ModuleBase` since the user has to call it for `DataBase` (and possibly other tensor the user created themselves).

> https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L97-L99
>   
>   Can this be replaced with Texar APIs? Quite a few hyperparameters hard-coded here

We have `get_optimizer` and `get_scheduler` under `texar.core.optimization`, but those functions are equally hard to use. We could move the magic numbers to a configuration file.

> https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L102
>   
>   Is it common / necessary ?

Not really. This only frees the CUDA memory cache, which contains blocks of memories that were previously allocated but not released by PyTorch. Keeping this cache prevents excessive memory allocation, and this cache is automatically managed by PyTorch. Normally, there's no need to manually empty the cache, unless in rare cases with high GPU memory usage.

> https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L114
>   
>   Is it common / necessary ?

Yes. This prevents PyTorch from computing gradients, which speeds things up by a large margin. This is also common practice. `torch.no_grad()` can also function as a function decorator.

>  https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/texar/modules/transformer.py#L33-L34
>   
>   Avoid magic constant

And actually, I think these should be loaded from the `Vocab` instance.

**Comment by haoransh** [comment]
> * [ ]  `texar.modules.Transformer` in its current form is far from being qualified as a Texar module. Please move it to under examples/

Okay. I will fix these issues.

> * [ ]  Can we use Texar data modules instead of torchtext which is not as clean as ours, e.g.,
>   https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L185-L192

As Zecong said, this is not supported right now. Our current implementation is a common practice to support dynamic batch with Pytorch.

> * [ ]  https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L43-L46
>   
>   The explanation "Either train_and_evaluate or test" is inconsistent to the code, which allows "train_and_evaluate"/"eval"/"test"....   Also, call it either "eval" or "evaluate", not both

Thanks, I will fix this.

> * [ ]  https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L73
>   
>   avoid using uncommon APIs like `getattr`, which harms readability. Is it necessary to use `getattr` here?

As zecong said, this is to avoid the condition when users don't specify the `beam_width` parameter explicitly in configuration file.

> * [ ]  https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L82-L86
>   
>   Is it necessary for all models? Can we simply this? E.g., add a `maybe_to_cuda()` method to `texar.ModuleBase`

I'm not very confident which is better. We may discuss the norm for all `examples`.

>   Is it common / necessary ?
> * [ ]  https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/transformer/transformer_main.py#L137
>   
>   If it's only for temporary file, use python `tempfile`

it's the plain-text outputs of the model. So I would prefer to write them to a file where the programmers can check the result externally ( probably they can use a listener to monitor the content of such files to inspect the model training ).

**Comment by haoransh** [comment]
> > https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/texar/modules/transformer.py#L33-L34
> > 
> > Avoid magic constant
> 
> And actually, I think these should be loaded from the `Vocab` instance.

This will be fixed soon. I will create a `Vocab` instance according to the recommended workflow in texar-pytorch.
